[GUI] QT 5.7

Opening an issue to track potential QT5.7 improvements, so that comments don't get lost on closed PRs.

QT 5.7 has been [released](https://blog.qt.io/blog/2016/06/16/qt-5-7-released/).

```
Starting with Qt 5.7, we will require a C++11 compliant compiler to build and use Qt. 
This allows us to use many of the new features in Qt itself, and renew our codebase 
using features such as constexpr, move semantics, range-for and auto. These efforts 
are well under way and will continue throughout the next versions.

#### Qt Core
- Qt uses poll() instead of select() in it’s event loop allowing for more than 1024 file 
descriptors
- Qt now uses std::atomic for it’s atomic classes
```

https://github.com/bitcoin/bitcoin/pull/8053#issuecomment-226714893

```
Though it's worth noting for a future pr: 5.7 improves a few build-related
things for us:
- allows for split native tools build.
- passes flags (like -D and -B only to the host toolchain)

(I'm on mobile so those are from memory, may not be 100% accurate. I also
haven't tested that those work as intended.)

The first should allow us to split up the build into more reasonable chunks
(native_qttools and qtlibs).

The second should allow us to drop the xcb patch in favor of using more
detailed flags. It should also allow us to drop the hackish env vars we use.
```

QT 5.7.0 is now the default qt5 install via [brew](https://github.com/Homebrew/homebrew-core/blob/master/Formula/qt5.rb) on OS X, it's currently failing compilation with "Qt requires C++11 support".
```
In file included from qt/macdockiconhandler.mm:5:
In file included from ./qt/macdockiconhandler.h:8:
In file included from /usr/local/Cellar/qt5/5.7.0/include/QtWidgets/QMainWindow:1:
In file included from /usr/local/Cellar/qt5/5.7.0/include/QtWidgets/qmainwindow.h:43:
In file included from /usr/local/Cellar/qt5/5.7.0/include/QtWidgets/qwidget.h:43:
In file included from /usr/local/Cellar/qt5/5.7.0/include/QtGui/qwindowdefs.h:43:
In file included from /usr/local/Cellar/qt5/5.7.0/include/QtCore/qglobal.h:1145:
In file included from /usr/local/Cellar/qt5/5.7.0/include/QtCore/qatomic.h:46:
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qbasicatomic.h:61:4: error: "Qt requires C++11 support"
#  error "Qt requires C++11 support"
   ^
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qbasicatomic.h:90:13: error: unknown type name 'QAtomicOps'
    typedef QAtomicOps<T> Ops;
            ^
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qbasicatomic.h:90:23: error: expected member name or ';' after
      declaration specifiers
    typedef QAtomicOps<T> Ops;
    ~~~~~~~~~~~~~~~~~~^
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qbasicatomic.h:93:23: error: use of undeclared identifier
      'QAtomicOpsSupport'
    Q_STATIC_ASSERT_X(QAtomicOpsSupport<sizeof(T)>::IsSupported, "template parameter is an integral of ...
                      ^
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qbasicatomic.h:93:53: error: no member named 'IsSupported' in the
      global namespace
    Q_STATIC_ASSERT_X(QAtomicOpsSupport<sizeof(T)>::IsSupported, "template parameter is an integral of ...
                                                  ~~^
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qglobal.h:761:63: note: expanded from macro 'Q_STATIC_ASSERT_X'
#define Q_STATIC_ASSERT_X(Condition, Message) Q_STATIC_ASSERT(Condition)
                                                              ^~~~~~~~~
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qglobal.h:756:110: note: expanded from macro 'Q_STATIC_ASSERT'
  ...__COUNTER__) = sizeof(QStaticAssertFailure<!!(Condition)>)}
                                                   ^~~~~~~~~
In file included from qt/macdockiconhandler.mm:5:
In file included from ./qt/macdockiconhandler.h:8:
In file included from /usr/local/Cellar/qt5/5.7.0/include/QtWidgets/QMainWindow:1:
In file included from /usr/local/Cellar/qt5/5.7.0/include/QtWidgets/qmainwindow.h:43:
In file included from /usr/local/Cellar/qt5/5.7.0/include/QtWidgets/qwidget.h:43:
In file included from /usr/local/Cellar/qt5/5.7.0/include/QtGui/qwindowdefs.h:43:
In file included from /usr/local/Cellar/qt5/5.7.0/include/QtCore/qglobal.h:1145:
In file included from /usr/local/Cellar/qt5/5.7.0/include/QtCore/qatomic.h:46:
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qbasicatomic.h:93:5: error: type name requires a specifier or
      qualifier
    Q_STATIC_ASSERT_X(QAtomicOpsSupport<sizeof(T)>::IsSupported, "template parameter is an integral of ...
    ^
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qglobal.h:761:47: note: expanded from macro 'Q_STATIC_ASSERT_X'
#define Q_STATIC_ASSERT_X(Condition, Message) Q_STATIC_ASSERT(Condition)
                                              ^
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qglobal.h:756:121: note: expanded from macro 'Q_STATIC_ASSERT'
  ...__COUNTER__) = sizeof(QStaticAssertFailure<!!(Condition)>)}
                                                              ^
In file included from qt/macdockiconhandler.mm:5:
In file included from ./qt/macdockiconhandler.h:8:
In file included from /usr/local/Cellar/qt5/5.7.0/include/QtWidgets/QMainWindow:1:
In file included from /usr/local/Cellar/qt5/5.7.0/include/QtWidgets/qmainwindow.h:43:
In file included from /usr/local/Cellar/qt5/5.7.0/include/QtWidgets/qwidget.h:43:
In file included from /usr/local/Cellar/qt5/5.7.0/include/QtGui/qwindowdefs.h:43:
In file included from /usr/local/Cellar/qt5/5.7.0/include/QtCore/qglobal.h:1145:
In file included from /usr/local/Cellar/qt5/5.7.0/include/QtCore/qatomic.h:46:
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qbasicatomic.h:95:14: error: use of undeclared identifier 'Ops'
    typename Ops::Type _q_value;
             ^
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qbasicatomic.h:95:19: error: expected a qualified name after
      'typename'
    typename Ops::Type _q_value;
                  ^
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qbasicatomic.h:95:23: error: expected ';' at end of declaration list
    typename Ops::Type _q_value;
                      ^
                      ;
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qbasicatomic.h:99:44: error: use of undeclared identifier 'Ops'
    T load() const Q_DECL_NOTHROW { return Ops::load(_q_value); }
                                           ^
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qbasicatomic.h:99:54: error: use of undeclared identifier '_q_value'
    T load() const Q_DECL_NOTHROW { return Ops::load(_q_value); }
                                                     ^
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qbasicatomic.h:100:45: error: use of undeclared identifier 'Ops'
    void store(T newValue) Q_DECL_NOTHROW { Ops::store(_q_value, newValue); }
                                            ^
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qbasicatomic.h:100:56: error: use of undeclared identifier '_q_value'
    void store(T newValue) Q_DECL_NOTHROW { Ops::store(_q_value, newValue); }
                                                       ^
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qbasicatomic.h:102:51: error: use of undeclared identifier 'Ops'
    T loadAcquire() const Q_DECL_NOTHROW { return Ops::loadAcquire(_q_value); }
                                                  ^
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qbasicatomic.h:102:68: error: use of undeclared identifier '_q_value'
    T loadAcquire() const Q_DECL_NOTHROW { return Ops::loadAcquire(_q_value); }
                                                                   ^
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qbasicatomic.h:103:52: error: use of undeclared identifier 'Ops'
    void storeRelease(T newValue) Q_DECL_NOTHROW { Ops::storeRelease(_q_value, newValue); }
                                                   ^
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qbasicatomic.h:103:70: error: use of undeclared identifier '_q_value'
    void storeRelease(T newValue) Q_DECL_NOTHROW { Ops::storeRelease(_q_value, newValue); }
                                                                     ^
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qbasicatomic.h:107:86: error: use of undeclared identifier 'Ops'
    static Q_DECL_CONSTEXPR bool isReferenceCountingNative() Q_DECL_NOTHROW { return Ops::isReferenceCou...
                                                                                     ^
/usr/local/Cellar/qt5/5.7.0/include/QtCore/qbasicatomic.h:108:88: error: use of undeclared identifier 'Ops'
    static Q_DECL_CONSTEXPR bool isReferenceCountingWaitFree() Q_DECL_NOTHROW { return Ops::isReferenceC...
                                                                                       ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
1 warning and 20 errors generated.
make[2]: *** [qt/qt_bitcoin_qt-macdockiconhandler.o] Error 1
```

@fanquake - How do you keep beating me to the punch? :) Was about to report this. My workaround for now is to use `brew switch qt5 5.5.1_2` to go back to 5.5, which seems to compile without issues. I don't know offhand if this is because I did a `brew tap homebrew/versions` earlier, which led to homebrew/versions/qt55 showing up when running `brew search qt5`.

On a somewhat related note, I'm curious what Core's policy is regarding warnings in code. OSX 10.12 turns up a lot of warnings in the LevelDB support code, such as the following.

```
./leveldb/port/atomic_pointer.h:55:3: warning: 'OSMemoryBarrier' is deprecated: first deprecated in macOS 10.12 - Use
      std::atomic_thread_fence() from <atomic> instead [-Wdeprecated-declarations]
  OSMemoryBarrier();
```

Different issue, I know, but I'm wondering if it's time to update LevelDB, assuming Google has fixed this issue. (I haven't compiled the upstream version.)

--- a/configure.ac
+++ b/configure.ac
@@ -45,7 +45,6 @@ else
   CXXFLAGS_overridden=no
 fi
 AC_PROG_CXX
-m4_ifdef([AC_PROG_OBJCXX],[AC_PROG_OBJCXX])
 
 dnl By default, libtool for mingw refuses to link static libs into a dll for
 dnl fear of mixing pic/non-pic objects, and import/export complications. Since
@@ -59,6 +58,15 @@ dnl Require C++11 compiler (no GNU extensions)
 AX_CXX_COMPILE_STDCXX([11], [noext], [mandatory])
 dnl Check if -latomic is required for <std::atomic>
 CHECK_ATOMIC
+
+dnl Unless the user specified OBJCXX, force it to be the same as CXX. This ensures
+dnl that we get the same -std flags for both.
+m4_ifdef([AC_PROG_OBJCXX],[
+if test "x${OBJCXX+set}" = "x"; then
+  OBJCXX="${CXX}"
+fi
+AC_PROG_OBJCXX
+])
 
 dnl Libtool init checks.
 LT_INIT([pic-only])

